### PR TITLE
[AMD][GLM-5.2] Keep GlmMoeDsa MoE e_score_correction_bias in fp32

### PR DESCRIPTION
**Pull request by xiaobochen-amd**
## Problem

GLM-5.2's MoE scores experts with `sigmoid(logits) + e_score_correction_bias` and takes the top 8.

That bias is fp32 in the checkpoint, with values in [6.817, 7.063] — a spread of 0.246 sitting up at 7. Near 7, bf16's smallest step is 0.03125, so it can only represent about 8 distinct values there. `MoEGate.__init__` casts the parameter to bf16 whenever a quant_config is present, so 238 distinct biases collapse into 8.

The result is that the wrong experts get picked. Routing the same real gating output both ways, 98.5% of tokens select a different top-8 set. HF keeps this tensor in fp32.

## Change

Two places have to change together:

1. `MoEGate.__init__`: skip the bf16 cast for GLM-5.2.
2. `biased_grouped_topk_gpu`, aiter path: it cast the bias back down before the kernel call. Keep an fp32 bias in fp32 and upcast the gating logits instead. This is gated on the bias dtype, not the architecture, so a model whose bias is already bf16 is unchanged.

## Cost

8k in / 1k out, MI355X, TP8, concurrency 32, two reps per arm:

| output tok/s | this PR | parent | delta |
|---|---|---|---|
| MTP off | 947.6 | 977.5 | -3.06% |
| MTP on | 1338.4 | 1323.7 | within noise |

3% slower without MTP. With MTP there is no measurable difference — the parent arm's own two reps differ by 18.1, more than the 14.7 between arms.

The overhead is per gating call, independent of how many tokens that forward produces. Without MTP one token carries all of it; with MTP the accept length is 5.13, so the same call is spread over ~5 tokens.

## Accuracy

Standard benchmarks cannot see this (GSM8K 0.941 → 0.947, GPQA-Diamond 0.8333 → 0.8182, both within error bars) — MoE experts are redundant, so the wrong expert often still gives the right answer. This is a routing-correctness fix, not a benchmark win.

---
### CI States

Latest PR Test (Base): :no_entry_sign: [Run #34112742506](https://github.com/sgl-project/sglang/actions/runs/34112742506)
Latest PR Test (Extra): :white_check_mark: [Run #34176132543](https://github.com/sgl-project/sglang/actions/runs/34176132543)
Latest PR Test (AMD ROCm 7.2): :x: [Run #34112742446](https://github.com/sgl-project/sglang/actions/runs/34112742446)

### CODE REVIEW

File: python/sglang/srt/models/deepseek_v2.py [hunk 1]
```diff
@@ -463,6 +463,13 @@ def forward(
         return x
 
 
+def _is_glm_moe_dsa(config) -> bool:
+    """True for GLM-5.2 GlmMoeDsa models -- the ``"GlmMoeDsa"`` substring matches
+    both the main arch ``GlmMoeDsaForCausalLM`` and the NextN draft head
+    ``GlmMoeDsaForCausalLMNextN`` (rewritten in ``configs/model_config.py``)."""
+    return any("GlmMoeDsa" in arch for arch in (config.architectures or []))
```
Comment (1am9trash) [inline python/sglang/srt/models/deepseek_v2.py]: Hi, @xiaobochen-amd 
Should we move `_is_glm_moe_dsa` to `python/sglang/srt/configs/model_config.py`? Similar arch checks (`is_deepseek_dsa`, `is_kimi_k3`) live there and are easier to reuse.

**Comment by 1am9trash** [comment]
Hi, @xiaobochen-amd 
You noted keeping the `MoEGate` bias in fp32 costs +4.5µs per gating call. Do you have e2e throughput numbers after the change (e.g. 8k/1k)? How much regression that actually costs?
Seem it's a tradeoff between acc and perf.

**Comment by 1am9trash** [comment]
image

Update branch for CI requirement.

**Comment by xiaobochen-amd** [comment]
Hi @1am9trash 

Ran 8k/1k on MI355X, TP8, concurrency 32. Two reps per arm, arms interleaved, this pr vs its parent (7e75115).

| output tok/s | this PR | parent | delta |
|---|---|---|---|
| MTP off | 947.6 | 977.5 | -3.06% |
| MTP on | 1338.4 | 1323.7 | within noise |

The 3% without MTP is real: the gap is 29.9 and the two reps within an arm differ by only 4.7.

With MTP there is nothing to measure. The parent arm's own two reps differ by 18.1, more than the 14.7 between arms, so that row is noise — I would not claim the PR is faster there.

Why the two rows differ: the overhead is per gating call, independent of how many tokens that forward produces. Without MTP one token carries all of it; with MTP the accept length is 5.13 and the same call is spread over ~5 tokens. Accept length is essentially unchanged between arms (5.134 vs 5.114), so the routing change does not affect draft acceptance either way.

Also a correction to the PR description: I wrote "about 1.2% of a decode step at most". Measured is 3.06%, so that number is wrong and I will fix it.

On the tradeoff framing: there is no accuracy gain to trade away — benchmarks cannot see this problem at all. It is routing correctness for 3% without MTP, and no measurable cost with MTP.

I will move `_is_glm_moe_dsa` to `configs/model_config.py` as you suggested, and update the branch.

File: python/sglang/srt/configs/model_config.py [hunk 2]
```diff
@@ -139,6 +139,15 @@ def is_deepseek_dsa(config) -> bool:
...
 
+def is_glm_moe_dsa(config) -> bool:
+    """True for GLM-5.2 GlmMoeDsa models. The ``"GlmMoeDsa"`` substring matches
+    both the main arch ``GlmMoeDsaForCausalLM`` and the NextN draft head
+    ``GlmMoeDsaForCausalLMNextN``."""
+    return any(
+        "GlmMoeDsa" in arch for arch in (_hf_attr(config, "architectures") or [])
+    )
+
+
```
Comment (HaiShaw) [inline python/sglang/srt/configs/model_config.py:150]: Can write this similar to following helpers, if to cover both "GlmMoeDsaForCausalLM" and "GlmMoeDsaForCausalLMNextN"?

Comment (xiaobochen-amd) [reply python/sglang/srt/configs/model_config.py:161]: Good call — switched to the explicit two-name form through `_hf_arch()`, matching
`is_deepseek_dsa` and `is_kimi_k3` right next to it. `is_deepseek_dsa` already
lists these same two architectures, so this keeps one idiom in the file and makes
the predicate greppable by arch name.

**Review by 1am9trash — APPROVED** [review]
Thanks for update. The comments are resolved.

**Review by HaiShaw — APPROVED** [review]

**Comment by HaiShaw** [comment]
AMD AITER specific changes only

**Comment by b8zhong** [comment]
Hi @xiaobochen-amd , I'm wondering if you could also help fix the router GEMM part of https://github.com/sgl-project/sglang/issues/34857? I think it might also cause some precision issues similar to this one.
